scripts/esbuild-commands.js

```diff
@@ -3,32 +3,51 @@
 const chalk = require('chalk')
 const esbuild = require('esbuild')
 
-const createEsbuildTask = (options = {}, { rebuildDelay, rebuildThrottle } = {}) => {
-  if (!options.plugins) options.plugins = []
+/**
+ * @param {Object} esbuildOptions - Native esbuild options.
+ * @param {Object} otherOptions - Additional options that are not part of the esbuild API.
+ * @returns {Object}
+ */
+const createEsbuildTask = (esbuildOptions = {}, otherOptions = {}) => {
+  if (!esbuildOptions.plugins) esbuildOptions.plugins = []
 
-  options.plugins[options.plugins.length] = defaultPlugin
+  // Make sure our 'default' plugin is always included.
+  esbuildOptions.plugins.push(defaultPlugin)
 
   return {
-    options,
+    options: { ...esbuildOptions, ...otherOptions },
 
     // Internal state.
     state: {
       // Holds the latest esbuild result object.
       result: null
     },
 
-    // Calls to run can be batched and throttled according to the corresponding options.
+    // Used to do an initial build.
     async run () {
-      const { options, state } = this
-
-      try {
-        if (state.result) {
-          await state.result.rebuild()
-        } else {
-          state.result = await esbuild.build(options)
-        }
-      } catch (error) {
-        console.log(error)
+      const { state } = this
+
+      if (state.result && esbuildOptions.incremental) {
+        throw new Error('Cannot use run() again on this task. Use rerun() instead.')
+      }
+      state.result = await esbuild.build(esbuildOptions)
+
+      if (otherOptions.postoperation) {
+        await otherOptions.postoperation()
+      }
+    },
+
+    // Used to do a rebuild after an input file event has been detected.
+    async rerun ({ fileEventName, filePath }) {
+      const { state } = this
+
+      if (!state.result || !esbuildOptions.incremental) {
+        throw new Error('Cannot use rerun() on this task. Use run() instead.')
+      }
+      await state.result.rebuild()
+
+      if (otherOptions.postoperation) {
+        await otherOptions.postoperation({ fileEventName, filePath })
```

Comment:
These two functions, `run` and `rerun`, are nearly identical except for the difference between:

```js
      state.result = await esbuild.build(esbuildOptions)
```

And:

```js
      await state.result.rebuild()
```

Couldn't therefore the logic be simplified so that there is only a single `run` function that keeps track of whether or not it has already been called before, and proceeds accordingly?

Reply:
There was indeed a single `run()` method before, which used `state.result` to check whether it had already been called.

Since additional arguments must now be accepted when a rebuild occurs, I've been considering whether `run()` should be split in two distinct methods with different signatures, rather than a single catch-all method which accepts optional arguments.
Besides, esbuild itself also has both `build()` and `rebuild()` methods.

It's a tradeoff between two simpler functions and a more complex one.
So I've tried this approach somewhat as an experiment, but I don't mind also trying the monolithic method.

Reply:
Both should be roughly equivalent in that respect at first guess (I might be wrong though).
In any case, issue #1085 will likely involve several refactorings.